### PR TITLE
Improve getHintedBundleGroups performance in JSRuntime

### PR DESCRIPTION
**Pull request by thebriando**
# ↪️ Pull Request

Improves the performance of the `getHintedBundleGroups` function in the JSRuntime by memoizing the async dependencies for each bundle after traversing.

## 💻 Examples

Before/after results of getHintedBundleGroups performance (before is on the left):
image

This also improves the performance of `applyRuntimes` overall since we warm the cache in `getHintedBundleGroups`, which can prevent another bundle traversal in `apply`.

image

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 5.41s | -115.00ms |
| Cached | 771.00ms | +24.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 31.00ms | -11.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 62.00ms | -10.00ms 🚀 |
| dist/legacy/index.6fa78f1c.js | 1.07kb | +0.00b | 1.20s | -78.00ms 🚀 |
| dist/legacy/index.afdbda6b.css | 77.00b | +0.00b | 1.09s | -60.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 58.00ms | -26.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 58.00ms | +4.00ms ⚠️ |
| dist/legacy/index.6fa78f1c.js | 1.07kb | +0.00b | 77.00ms | -7.00ms 🚀 |
| dist/modern/index.df30fe40.js | 1.07kb | +0.00b | 78.00ms | -6.00ms 🚀 |
| dist/legacy/index.html | 701.00b | +0.00b | 77.00ms | -7.00ms 🚀 |
| dist/modern/index.html | 701.00b | +0.00b | 77.00ms | -7.00ms 🚀 |
| dist/legacy/index.afdbda6b.css | 77.00b | +0.00b | 77.00ms | -7.00ms 🚀 |
| dist/modern/index.ef3960f9.css | 77.00b | +0.00b | 78.00ms | -6.00ms 🚀 |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 23.68s | +121.00ms |
| Cached | 12.51s | -94.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/NotFound.141e0b29.js | 532.00b | +0.00b | 271.00ms | +15.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 479.93kb | +0.00b | 53.00ms | -8.00ms 🚀 |
| dist/PermalinkedComment.a5120a24.js | 4.22kb | +0.00b | 53.00ms | -6.00ms 🚀 |
| dist/UserProfile.95e14af2.js | 1.70kb | +0.00b | 53.00ms | -6.00ms 🚀 |
| dist/NotFound.141e0b29.js | 532.00b | +0.00b | 37.00ms | -22.00ms 🚀 |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 37.00ms | -7.00ms 🚀 |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.22m | -2.67s |
| Cached | 2.89s | -150.00ms 🚀 |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/EmojiPickerComponent.f8cd332f.js | 139.08kb | +0.00b | 10.28s | +563.00ms ⚠️ |
| dist/card.93c390cd.js | 53.76kb | +0.00b | 1.22m | +1.04m ⚠️ |
| dist/DatePicker.cf283b8a.js | 21.16kb | +0.00b | 10.25s | +922.00ms ⚠️ |
| dist/ui.dc2b3f59.js | 14.11kb | +0.00b | 1.22m | +1.04m ⚠️ |
| dist/workerHasher.1a860446.js | 11.90kb | +0.00b | 8.59s | -1.08m 🚀 |
| dist/component.aaadd915.js | 6.27kb | +0.00b | 6.17s | +876.00ms ⚠️ |
| dist/EmojiPickerComponent.53446e07.js | 3.68kb | +0.00b | 8.57s | -1.78s 🚀 |
| dist/16.4fc0f868.js | 2.49kb | +0.00b | 2.31s | +144.00ms ⚠️ |
| dist/ResourcedEmojiComponent.87af091a.js | 2.15kb | +0.00b | 36.13s | +25.77s ⚠️ |
| dist/date.7c1c4784.js | 1.96kb | +0.00b | 5.49s | -1.44s 🚀 |
| dist/feedback.89b37c2d.js | 1.86kb | +0.00b | 6.01s | -1.94s 🚀 |
| dist/16.fb8534c0.js | 1.79kb | +0.00b | 6.48s | +896.00ms ⚠️ |
| dist/workerHasher.209d7f04.js | 1.75kb | +0.00b | 10.28s | -1.05m 🚀 |
| dist/16.34fd12e7.js | 1.46kb | +0.00b | 6.17s | +875.00ms ⚠️ |
| dist/expand.a88194c2.js | 1.38kb | +0.00b | 10.25s | +2.30s ⚠️ |
| dist/16.867d534c.js | 1.26kb | +0.00b | 6.17s | +876.00ms ⚠️ |
| dist/divider.6089d072.js | 1.25kb | +0.00b | 5.50s | -1.44s 🚀 |
| dist/component.0a97dc28.js | 1.23kb | +0.00b | 2.31s | +145.00ms ⚠️ |
| dist/16.a43add3b.js | 1.23kb | +0.00b | 2.32s | +146.00ms ⚠️ |
| dist/decision.0a6d7582.js | 1.21kb | +0.00b | 5.49s | -1.44s 🚀 |
| dist/heading1.fcd21443.js | 1.18kb | +0.00b | 5.85s | -31.16s 🚀 |
| dist/media-card-analytics-error-boundary.5602bdf2.js | 1.05kb | +0.00b | 1.22m | +1.02m ⚠️ |
| dist/simpleHasher.03c3cf37.js | 755.00b | +0.00b | 8.57s | -1.08m 🚀 |
| dist/index.html | 119.00b | +0.00b | 3.62s | +795.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.e3f5aa90.js | 2.32mb | +0.00b | 476.00ms | -68.00ms 🚀 |
| dist/pdfRenderer.705aa8d6.js | 1.11mb | -6.00b 🚀 | 599.00ms | -67.00ms 🚀 |
| dist/editorView.2c405cb8.js | 587.39kb | +0.00b | 618.00ms | -80.00ms 🚀 |
| dist/popup.4af16ddf.js | 169.21kb | +0.00b | 632.00ms | -72.00ms 🚀 |
| dist/EmojiPickerComponent.f8cd332f.js | 139.08kb | +0.00b | 583.00ms | -71.00ms 🚀 |
| dist/Toolbar.a1a51fa7.js | 99.13kb | +0.00b | 632.00ms | -66.00ms 🚀 |
| dist/media-viewer.b6f54158.js | 72.60kb | +0.00b | 599.00ms | -67.00ms 🚀 |
| dist/card.a5c85477.js | 53.89kb | +0.00b | 599.00ms | -68.00ms 🚀 |
| dist/card.00dc27ce.js | 51.65kb | +0.00b | 632.00ms | -75.00ms 🚀 |
| dist/Modal.8e9ccd89.js | 38.79kb | +0.00b | 480.00ms | -85.00ms 🚀 |
| dist/esm.621f7fcd.js | 27.71kb | +0.00b | 583.00ms | -75.00ms 🚀 |
| dist/DatePicker.5bd26c52.js | 21.03kb | +0.00b | 583.00ms | -55.00ms 🚀 |
| dist/smartMediaEditor.c1fb8779.js | 16.69kb | +0.00b | 618.00ms | -80.00ms 🚀 |
| dist/dropzone.7ab612aa.js | 15.96kb | +0.00b | 607.00ms | -76.00ms 🚀 |
| dist/ui.dc2b3f59.js | 14.11kb | +0.00b | 599.00ms | -53.00ms 🚀 |
| dist/workerHasher.1a860446.js | 11.90kb | +0.00b | 584.00ms | -69.00ms 🚀 |
| dist/card.3fd8a427.js | 5.71kb | +0.00b | 599.00ms | -67.00ms 🚀 |
| dist/media-viewer.b91b6ee1.js | 4.01kb | +0.00b | 599.00ms | -67.00ms 🚀 |
| dist/EmojiPickerComponent.8447f1ef.js | 3.56kb | +0.00b | 599.00ms | -54.00ms 🚀 |
| dist/index.cfd8f78a.css | 3.46kb | +0.00b | 634.00ms | -78.00ms 🚀 |
| dist/dropzone.dcab7b1b.js | 3.44kb | +0.00b | 632.00ms | -51.00ms 🚀 |
| dist/clipboard.a5615679.js | 2.97kb | +0.00b | 607.00ms | -76.00ms 🚀 |
| dist/16.4fc0f868.js | 2.49kb | +0.00b | 476.00ms | -68.00ms 🚀 |
| dist/ResourcedEmojiComponent.87af091a.js | 2.15kb | +0.00b | 599.00ms | -54.00ms 🚀 |
| dist/card.bfb3f5f8.js | 2.15kb | +0.00b | 632.00ms | -51.00ms 🚀 |
| dist/date.7c1c4784.js | 1.96kb | +0.00b | 525.00ms | -51.00ms 🚀 |
| dist/images.c026df2d.js | 1.90kb | +0.00b | 535.00ms | -71.00ms 🚀 |
| dist/feedback.89b37c2d.js | 1.86kb | +0.00b | 571.00ms | -66.00ms 🚀 |
| dist/browser.dc14c9f2.js | 1.82kb | +0.00b | 607.00ms | -76.00ms 🚀 |
| dist/workerHasher.209d7f04.js | 1.75kb | +0.00b | 584.00ms | -74.00ms 🚀 |
| dist/workerHasher.b62cf8f4.js | 1.75kb | +0.00b | 618.00ms | -80.00ms 🚀 |
| dist/list-number.f10dae12.js | 1.68kb | +0.00b | 548.00ms | -61.00ms 🚀 |
| dist/status.e89e15e9.js | 1.68kb | +0.00b | 563.00ms | -56.00ms 🚀 |
| dist/code.17026d12.js | 1.61kb | +0.00b | 525.00ms | -51.00ms 🚀 |
| dist/link.5063f523.js | 1.53kb | +0.00b | 537.00ms | -69.00ms 🚀 |
| dist/heading6.95167c98.js | 1.53kb | +0.00b | 571.00ms | -66.00ms 🚀 |
| dist/heading3.59235a88.js | 1.51kb | +0.00b | 563.00ms | -79.00ms 🚀 |
| dist/16.85df7a6e.js | 1.51kb | +0.00b | 494.00ms | -29.00ms 🚀 |
| dist/emoji.0ed2a831.js | 1.45kb | +0.00b | 535.00ms | -71.00ms 🚀 |
| dist/heading5.83973595.js | 1.40kb | +0.00b | 571.00ms | -66.00ms 🚀 |
| dist/expand.a88194c2.js | 1.38kb | +0.00b | 571.00ms | -66.00ms 🚀 |
| dist/heading2.27558830.js | 1.33kb | +0.00b | 563.00ms | -74.00ms 🚀 |
| dist/16.0ef0d0ed.js | 1.31kb | +0.00b | 525.00ms | -51.00ms 🚀 |
| dist/media-card-analytics-error-boundary.9f8e76f5.js | 1.30kb | +0.00b | 632.00ms | -72.00ms 🚀 |
| dist/mention.f75f6880.js | 1.29kb | +0.00b | 548.00ms | -62.00ms 🚀 |
| dist/heading4.00adf9a0.js | 1.29kb | +0.00b | 563.00ms | -79.00ms 🚀 |
| dist/layout.8d45cad7.js | 1.27kb | +0.00b | 535.00ms | -71.00ms 🚀 |
| dist/16.1e3b9f3d.js | 1.27kb | +0.00b | 494.00ms | -29.00ms 🚀 |
| dist/16.a86c2eba.js | 1.26kb | +0.00b | 494.00ms | -29.00ms 🚀 |
| dist/16.8e47d46b.js | 1.26kb | +0.00b | 494.00ms | -29.00ms 🚀 |
| dist/divider.6089d072.js | 1.25kb | +0.00b | 535.00ms | -71.00ms 🚀 |
| dist/quote.f0f2e340.js | 1.25kb | +0.00b | 558.00ms | -61.00ms 🚀 |
| dist/component.0a97dc28.js | 1.23kb | +0.00b | 476.00ms | -68.00ms 🚀 |
| dist/action.946aff12.js | 1.23kb | +0.00b | 525.00ms | -51.00ms 🚀 |
| dist/16.a43add3b.js | 1.23kb | +0.00b | 476.00ms | -68.00ms 🚀 |
| dist/decision.0a6d7582.js | 1.21kb | +0.00b | 528.00ms | -48.00ms 🚀 |
| dist/panel-warning.13de0cff.js | 1.21kb | +0.00b | 558.00ms | -61.00ms 🚀 |
| dist/list.36f0ec8a.js | 1.18kb | +0.00b | 548.00ms | -62.00ms 🚀 |
| dist/heading1.fcd21443.js | 1.18kb | +0.00b | 563.00ms | -74.00ms 🚀 |
| dist/panel-error.36583145.js | 1.11kb | +0.00b | 549.00ms | -70.00ms 🚀 |
| dist/panel.529e9757.js | 1.10kb | +0.00b | 558.00ms | -61.00ms 🚀 |
| dist/table.402ee877.js | 1.09kb | +0.00b | 563.00ms | -70.00ms 🚀 |
| dist/media-picker-analytics-error-boundary.807445d0.js | 1.05kb | +0.00b | 607.00ms | -76.00ms 🚀 |
| dist/panel-success.83756958.js | 1.05kb | +0.00b | 558.00ms | -61.00ms 🚀 |
| dist/media-card-analytics-error-boundary.5602bdf2.js | 1.05kb | +0.00b | 598.00ms | -69.00ms 🚀 |
| dist/media-card-analytics-error-boundary.fa8679e7.js | 1.05kb | +0.00b | 612.00ms | -96.00ms 🚀 |
| dist/panel-note.5ebd6d66.js | 1.05kb | +0.00b | 549.00ms | -70.00ms 🚀 |
| dist/media-viewer-analytics-error-boundary.bc280d38.js | 995.00b | +0.00b | 603.00ms | -63.00ms 🚀 |
| dist/simpleHasher.03c3cf37.js | 755.00b | +0.00b | 584.00ms | -69.00ms 🚀 |
| dist/simpleHasher.566203bd.js | 755.00b | +0.00b | 618.00ms | -80.00ms 🚀 |

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 16.36s | +137.00ms |
| Cached | 704.00ms | +36.00ms ⚠️ |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/Three.js | 580.68kb | +0.00b | 113.00ms | +13.00ms ⚠️ |

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/285025252397810184)**

**Comment by wbinnssmith** [comment]
My only concern with this approach is the cached values becoming inaccurate when bundle content changes, which can happen on subsequent builds, and even once the runtime results are merged into the main graph. We should add an integration test that adds new async dependencies and verifies that they're accounted for on subsequent builds.

@devongovett what do you think?

File: packages/runtimes/js/src/JSRuntime.js [hunk 1]
```diff
@@ -64,6 +65,14 @@ function getLoaders(
   return null;
 }
 
+let bundleAsyncDependencies = new WeakMap<
```
Comment (wbinnssmith) [inline packages/runtimes/js/src/JSRuntime.js:68]: This isn't the dependencies themselves, it's their resolutions. If we cache the dependencies themselves, we can even warm this cache during the initial traversal when determining them for loaders here:

https://github.com/parcel-bundler/parcel/blob/a4cf3e958ea1136b4829eb93615fba5328213471/packages/runtimes/js/src/JSRuntime.js#L82

**Review by wbinnssmith — APPROVED** [review]

File: packages/runtimes/js/src/JSRuntime.js [hunk 2]
```diff
@@ -79,18 +89,29 @@ export default (new Runtime({
...
+    let cachedDependencies = bundleDependencies.get(bundle);
 
-      let dependency = node.value;
-      if (dependency.isAsync && !dependency.isURL) {
-        asyncDependencies.push(dependency);
-      } else {
-        otherDependencies.push(dependency);
-      }
-    });
+    if (
```
Comment (wbinnssmith) [inline packages/runtimes/js/src/JSRuntime.js:94]: determining these could be moved into a function (e.g. `getDependencies`) and shared with the logic below in `getHintedBundleGroups`

**Review by mischnic — APPROVED** [review]
We did talk about making runtimes run once per bundle graph, then that map could be moved inside of the plugin method itself. But I think it's fine for now since runtimes need to be overhauled anyway.

**Pull request merged**